### DISCUSSION TITLE
Set bun version in bunfig file

### DISCUSSION THREAD

**Discussion by ftzi** [oven-sh/bun#10955] [Ideas]
I already had a situation in which, due to a mismatch in the bun's version used in CI and local bun, some errors were occurring in the lock file.

While in the CI we can specify the bun version, apparently there is no way to enforce project's users to use a specific bun version.

Can we have a field in bunfig that if `bun -v` is not that version, it would automatically install that version when `bun i` is run?

This would make upgrades to bun in our setup easier as we wouldn't need to tell the devs to upgrade bun.

**Comment by karlhorky** [comment]
Currently, there is:

1. `.bun-version`, a convention similar to .node-version - used to specify the Bun version
   - Official Bun GitHub Actions action https://github.com/oven-sh/setup-bun#using-version-file
   - Renovate bot https://github.com/renovatebot/renovate/issues/29941
2. [`packageManager` field in `package.json`](https://nodejs.org/api/packages.html#packagemanager) - Corepack will refuse to run with a different package manager using this field

So those 2 options look like better starting points than a new place in `bunfig.toml`

> there is no way to enforce project's users to use a specific bun version

However, I'm not sure whether either of these options above currently enforce the Bun version.

**Reply by charpeni** [reply]
Both options mentioned above are currently (as of Bun `1.2.22`) not enforcing a Bun version.

Corepack doesn't support Bun, and therefore the `packageManager` field in `package.json` has no effect (versus npm, pnpm, or yarn).

`.bun-version` is quite convenient for GitHub Actions, but there's nothing locally to check or even enforce the specified version. I presumed there's an issue describing this feature somewhere?

Regardless, I totally discarded the `packageManager` field for Bun and went with `.bun-version` combined with a custom `preinstall` script as such:

```ts
/**
 * Check if the installed Bun version matches the required version as defined in `.bun-version`.
 *
 * As of today (September 2025), Corepack doesn't support Bun as a valid package manager.
 */
(async () => {
	const required = await Bun.file(".bun-version").text();
	const actual = Bun.version;

	if (actual.trim() !== required.trim()) {
		console.error(`Error: Bun version ${required.trim()} required, but ${actual} is installed.`);
		process.exit(1);
	}
})();
```

This brought me closer to what a setup with pnpm would have been. Hoping it could help someone else too. 🙂

**Reply by drake-nathan** [reply]
Thank you, this was incredibly helpful. Since I use Sherif as a monorepo check, it requires the `packageManager` field, and I prefer that method anyways. So here's how you can tweak that script to use the `package.json` instead of `.bun-version`. This way, you simply remove the pre-install if/when corepack adds bun support.

Then you'll set `bun-version-file: "package.json"` in your CI.

```
/**
 * Check if the installed Bun version matches the required version as defined in package.json's packageManager field.
 * As of today (September 2025), Corepack doesn't support Bun as a valid package manager.
 */
(async () => {
  const packageJson = await Bun.file("package.json").json();
  const packageManager = packageJson.packageManager;

  if (!packageManager) {
    console.error("Error: No packageManager field found in package.json");
    process.exit(1);
  }

  // Parse "bun@x.x.x" format
  const match = packageManager.match(/^bun@(?<version>.+)$/);

  if (!match) {
    console.error(
      `Error: packageManager field "${packageManager}" is not in the format "bun@x.x.x"`,
    );
    process.exit(1);
  }

  const required = match.groups.version;
  const actual = Bun.version;

  if (actual.trim() !== required.trim()) {
    console.error(
      `Error: Bun version ${required.trim()} required, but ${actual} is installed.`,
    );
    process.exit(1);
  }
})();
```

**Reply by charpeni** [reply]
In the next version of [`setup-bun`](https://github.com/oven-sh/setup-bun) (merged, but not yet released), it will support reading the version from `packageManager` field out of the box, so the `.bun-version` won’t be required anymore! 🙂
